### ISSUE TITLE
Copy & Paste issue in vscode on mac

### ISSUE THREAD

**Issue by jhuesser** [hediet/vscode-drawio#28]
Hi there,

We've tested this on 2 different MacBooks with vscode Version:
`Version: 1.45.0
Commit: d69a79b73808559a91206d73d7717ff5f798f23c
Date: 2020-05-07T15:57:33.467Z
Electron: 7.2.4
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Darwin x64 19.4.0`

Using copy & paste from the context menu works fine, but there I can't see a shortcut for paste and cmd+v doesn't works:
![Imgur](https://i.imgur.com/tB3YVn0.png)

Also cmd+c to copy a shape also doesn't work. If I select a shape press cmd+c right click -> paste here, it pastes the shape that was copied with right click -> copy.

**Comment by hediet** [comment]
So to understand you correctly:
* Copying from the context menu works always.
* Pasting from the context menu works always.
* Copying from the context menu and pasting using cmd+v (or ctrl/fn+v?) does not do what it should do.
* Copying with cmd+c (or ctrl/fn+c) and pasting using the context menu does not do what it should do.
* And then, clearly, copying and pasting both using shortcuts does not work.

So basically shortcuts don't work at all?
I don't have a mac and cannot really debug it as they don't offer a VM for windows :(

**Comment by jhuesser** [comment]
Hi,

Yes that is correct. Other shortcuts, for example cmd+z and cmd+a for undo / select all works fine.

I've also checked cmd+x for cut right now, and it doesn't work. So yes basically shortcuts for cut, copy and paste doesn't work.

**Comment by hediet** [comment]
It would be awesome if someone who is using a Mac could investigate this further. It works on Windows though.
I have no clue how to fix this and without being able to reproduce it on Windows, I cannot really debug it.

**Comment by TTSKarlsson** [comment]
I'm not using a mac, but if it helps MDN https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState states that the command key is called "Meta".

**Comment by Speedy37** [comment]
draw.io supports Mac without trouble, it seams, in the context of vscode on MacOS events aren't captured at all.

I added `window.addEventListener("paste", () => log("paste"))`, nothing is printed.
It does print on windows.

No idea why this happens only on MacOS.

**Comment by hediet** [comment]
@Speedy37 so you think there is just no way to implement it in my extension?

**Comment by Speedy37** [comment]
I think, we need to know the root cause.

Awesome extension btw.

**Comment by Speedy37** [comment]
The "paste" event is working in the "WebView > iframe"
It's not working inside the draw.io iframe "WebView > iframe > iframe"

**Comment by hediet** [comment]
Thanks for your research, @Speedy37!
It is not so easy to get rid of the second, inner iframe (but it might be possible). Do you know whether the event can be forwarded into the inner iframe in the meantime?

**Comment by Speedy37** [comment]
I will allocate some time to fix this later this week. Either by:

- working around it by forwarding the event (should be easy to do)
- removing the iframe (this requires fixing draw.io as the first iframe is without a uri, making some web api locked)
- finding the origin of the bug in electron

**Comment by hediet** [comment]
> working around it by forwarding the event (should be easy to do)

I too think this might be the easiest way.

> removing the iframe (this requires fixing draw.io as the first iframe is without a uri, making some web api locked)

I already have some ideas. I opened the ticket #42 for that.

**Comment by Speedy37** [comment]
I saw you did some work, you packaged some files that aren't present in git.
I used it as starting point to patch draw.io to start directly. It does work, but requires a few patches.

**Comment by jhuesser** [comment]
Hi,

Just a follow up: cut, copy & paste now works perfectly on our Macs. Thank you very much!

**Comment by hediet** [comment]
Thanks for the follow up ;) I'm glad it works!

**Comment by Kaijun** [comment]
@jhuesser @hediet i'm also encountering the same problem. 

in which release does this problem get solved? 

i'm on v0.7.2 btw.

**Comment by Kaijun** [comment]
ok it seems to be a problem with drawio, which also happened to drawio-desktop: https://github.com/jgraph/drawio-desktop/issues/358

**Comment by hediet** [comment]
Do you think that is something that should be fixed in this extension? Or in VS Code or in Drawio?

**Comment by Kaijun** [comment]
> Do you think that is something that should be fixed in this extension? Or in VS Code or in Drawio?

@hediet i'm not pretty sure for now since it happened randomly in drawio-desktop, but in vscode-drawio copy&paste doesn't work 100% in all cases

**Comment by jhuesser** [comment]
@Kaijun We are using v0.7.2 of the extension it for us it works. Since updating, we have never seen this issue occurring again.

**Comment by hediet** [comment]
@Kaijun do you experience the same problem with the hex editor or other web views?
Did you try out vscode insider?

**Comment by Kaijun** [comment]
@hediet it seems to be a bug in electron: https://github.com/jgraph/drawio-desktop/issues/2
